File: native/ios/SpreadDex/Appraise.swift
import Foundation

enum Appraise {
    static let skip = try! NSRegularExpression(
        pattern: #"\b(lots?\b|\d+\s+cards\b|bulk|binder|choose your|etb|booster|proxy|custom\b|plush|hoodie)\b"#,
        options: .caseInsensitive
    )

    static func skipListing(_ title: String) -> Bool {
        let r = NSRange(title.startIndex..., in: title)
        return skip.firstMatch(in: title, range: r) != nil
    }

    static func parseMoney(_ raw: String?) -> Double? {
        guard let raw, let n = Double(raw.replacingOccurrences(of: ",", with: "")) , n > 0 else { return nil }
        return n
    }

    static func tokens(_ query: String) -> [String] {
        let generic: Set<String> = ["pokemon", "pokémon", "tcg", "card", "cards", "holo", "rare"]
        return query.lowercased()
            .replacingOccurrences(of: "[^a-z0-9\\s]", with: " ", options: .regularExpression)
            .split(separator: " ").map(String.init)
            .filter { $0.count > 2 && !generic.contains($0) }
    }

    static func titleMatches(_ title: String, _ query: String) -> Bool {
        guard let t = tokens(query).first else { return true }
        return title.lowercased().contains(t)
    }

    static func detectGrade(_ text: String) -> String {
        let t = text.lowercased()
        if t.contains("psa 10") || t.contains("psa10") { return "PSA 10" }
        if t.contains("bgs 10") { return "BGS 10" }
        if t.contains("cgc 10") { return "CGC 10" }
        if t.contains("psa 9") { return "PSA 9" }
        return "raw"
    }

    static func conditionMult(_ text: String) -> Double {
        let t = text.lowercased()
        if t.contains("dmg") || t.contains("damaged") { return 0.2 }
        if t.contains("hp") { return 0.35 }
        if t.contains("mp") || t.contains("played") { return 0.55 }
        if t.contains("lp") { return 0.8 }
        return 1
    }

    static func gradeMult(card: TcgCard?, grade: String) -> Double {
        if grade == "raw" { return 1 }
        let set = "\(card?.setId ?? "") \(card?.setName ?? "")".lowercased()
        let vintage = set.range(of: "base1|jungle|fossil|neo|wotc|base set", options: .regularExpression) != nil
        if grade.contains("10") && !grade.contains("9.5") {
            if vintage { return 4 }
            return 1.25
        }
        return 1.35
    }

    static func nameQuery(_ title: String) -> String {
        let stop: Set<String> = ["pokemon", "pokémon", "tcg", "card", "cards", "holo", "rare", "nm", "psa", "bgs", "cgc", "english"]
        let cleaned = title.lowercased()
            .replacingOccurrences(of: #"https?://\S+"#, with: " ", options: .regularExpression)
            .replacingOccurrences(of: #"\$[0-9,.]+"#, with: " ", options: .regularExpression)
            .replacingOccurrences(of: #"[^a-z0-9'\s-]"#, with: " ", options: .regularExpression)
        let parts = cleaned.split(separator: " ").map(String.init)
            .filter { !$0.isEmpty && !stop.contains($0) && Int($0) == nil }
        return parts.prefix(4).joined(separator: " ").isEmpty ? "pokemon" : parts.prefix(4).joined(separator: " ")
    }

    static func appraise(card: TcgCard, listing: LiveListing, grade: String, extra: Double? = nil) -> Appraisal {
        let seed = [card.finishes.first(where: { $0.market != nil })?.market, extra].compactMap { $0 }
        let raw = seed.isEmpty ? nil : seed.reduce(0, +) / Double(seed.count)
        let adjusted = raw.map { $0 * conditionMult(listing.title) * gradeMult(card: card, grade: grade) }
        let allIn = (listing.price ?? 0) + listing.shipping
        let spread = (adjusted != nil && adjusted! > 0) ? (adjusted! - allIn) / adjusted! : nil
        let verdict: String
        switch spread {
        case .some(let s) where s >= 0.30: verdict = "steal"
        case .some(let s) where s >= 0.12: verdict = "good"
        case .some(let s) where s >= -0.08: verdict = "fair"
        case .some(let s) where s >= -0.30: verdict = "high"
        case .some: verdict = "avoid"
        default: verdict = "fair"
        }
        return Appraisal(market: raw, adjusted: adjusted, allIn: allIn, spread: spread, verdict: verdict)
    }
}

File: native/android/app/src/main/java/me/grok/dealdex/data/Appraise.kt
package me.grok.dealdex.data

object Appraise {
    private val skip = Regex(
        """\b(lots?\b|\d+\s+cards\b|bulk|binder|choose your|pick your|etb|booster|code card|proxy|custom\b|plush|hoodie|t-?shirt)\b""",
        RegexOption.IGNORE_CASE,
    )

    fun skipListing(title: String) = skip.containsMatchIn(title)

    fun parseMoney(raw: String?): Double? {
        if (raw.isNullOrBlank()) return null
        val n = raw.replace(",", "").toDoubleOrNull() ?: return null
        return if (n > 0) n else null
    }

    fun significantTokens(query: String): List<String> {
        val generic = setOf("pokemon", "pokémon", "tcg", "card", "cards", "holo", "rare")
        return query.lowercase().replace(Regex("[^a-z0-9\\s]"), " ")
            .split(Regex("\\s+")).filter { it.length > 2 && it !in generic }
    }

    fun titleMatches(title: String, query: String): Boolean {
        val t = significantTokens(query).firstOrNull() ?: return true
        return title.lowercase().contains(t)
    }

    /**
     * Kept in step with `detectGrade` in `src/lib/tcg/parse-listing.ts` so the
     * phone and the website reach the same verdict on the same listing.  PSA 8,
     * BGS 9.5, CGC 9.5 and ACE 10 used to fall through to "raw".
     */
    fun detectGrade(text: String): String {
        val t = text.uppercase().replace(Regex("\\s+"), " ")
        val checks = listOf(
            Regex("\\bPSA ?10\\b") to "PSA 10",
            Regex("\\bPSA ?9\\b") to "PSA 9",
            Regex("\\bPSA ?8\\b") to "PSA 8",
            Regex("\\bBGS ?10\\b") to "BGS 10",
            Regex("\\bBGS ?9\\.5\\b") to "BGS 9.5",
            Regex("\\bCGC ?10\\b") to "CGC 10",
            Regex("\\bCGC ?9\\.5\\b") to "CGC 9.5",
            Regex("\\bACE ?10\\b") to "ACE 10",
        )
        for ((re, grade) in checks) if (re.containsMatchIn(t)) return grade
        return "raw"
    }

    /**
     * A standalone `HP` in a Pokemon listing title is the hit-point stat, not
     * Heavily Played.  `"hp" in t` matched a large share of every scan and cut
     * the book to 35%, so the app called real deals overpriced.  `"lp" in t`
     * matched "Delphox"; `"mp" in t` matched "champion"; `"played" in t` even
     * matched "lightly played" before the LP branch could see it.
     */
    fun hasConditionCode(text: String, code: String): Boolean {
        // Bracketed, slash-joined or explicitly labelled: never a stat line.
        if (Regex("([(\\[/]|cond(ition)?[ :-]*)\\s*$code\\b", RegexOption.IGNORE_CASE).containsMatchIn(text)) {
            return true
        }
        val re = Regex("(^|[^a-z0-9])$code([^a-z0-9]|$)", RegexOption.IGNORE_CASE)
        for (m in re.findAll(text)) {
            val before = text.substring(0, m.range.first + m.groupValues[1].length)
            val after = text.substring(minOf(m.range.last + 1, text.length))
            if (Regex("\\d\\s*\\W?\\s*$").containsMatchIn(before)) continue
            if (Regex("^\\W?\\s*\\d").containsMatchIn(after)) continue
            return true
        }
        return false
    }

    fun detectCondition(text: String): Double {
        val t = text.lowercase()
        // Spelled out is unambiguous and wins.
        if ("damaged" in t || "poor condition" in t) return 0.2
        if ("heavily played" in t || "heavy play" in t) return 0.35
        if ("moderately played" in t || "moderate play" in t) return 0.55
        if ("lightly played" in t || "light play" in t) return 0.8
        if ("near mint" in t || "mint condition" in t) return 1.0

        if (hasConditionCode(t, "dmg")) return 0.2
        if (hasConditionCode(t, "hp")) return 0.35
        if (hasConditionCode(t, "mp")) return 0.55
        if (hasConditionCode(t, "lp")) return 0.8
        return 1.0
    }

    /** Mirrors `GRADE_MULT` / `PSA10_BY_BUCKET` in `src/lib/tcg/appraise.ts`. */
    private val gradeMultBase = mapOf(
        "PSA 10" to 2.8, "PSA 9" to 1.35, "PSA 8" to 0.95,
        "BGS 10" to 3.2, "BGS 9.5" to 1.8,
        "CGC 10" to 2.4, "CGC 9.5" to 1.4,
        "ACE 10" to 2.2,
    )

    fun gradeMult(card: TcgCard?, grade: String): Double {
        if (grade == "raw") return 1.0
        val base = gradeMultBase[grade] ?: 1.0
        if (card == null || !grade.endsWith("10")) return base
        val set = "${card.setId} ${card.setName}".lowercase()
        val rarity = card.rarity.orEmpty().lowercase()
        // Same vintage set list as the web: base2/base3/base4/gym/team rocket
        // were missing here, so those sets graded as modern.
        val vintage = Regex("base1|base2|base3|base4|jungle|fossil|neo|gym|team rocket|wotc")
            .containsMatchIn(set)
        val chase = Regex("illustration rare|special illustration|alt art|hyper rare|secret")
            .containsMatchIn(rarity)
        val bucket = when {
            vintage && rarity.contains("holo") -> 8.0
            vintage -> 4.0
            chase -> 2.1
            else -> 1.25
        }
        // Keep each 10-grade's own ratio to PSA 10 instead of flattening them.
        return bucket * (base / 2.8)
    }

    fun pickMarket(card: TcgCard): Double? =
        card.finishes.firstOrNull { it.market != null }?.market

    fun appraise(card: TcgCard, listing: LiveListing, grade: String, extra: Double? = null): Appraisal {
        val seed = listOfNotNull(pickMarket(card), extra)
        val raw = if (seed.isEmpty()) null else seed.average()
        val adjusted = raw?.let { it * detectCondition(listing.title) * gradeMult(card, grade) }
        val allIn = (listing.price ?: 0.0) + listing.shipping
        val spread = if (adjusted != null && adjusted > 0) (adjusted - allIn) / adjusted else null
        val verdict = when {
            spread == null -> "fair"
            spread >= 0.30 -> "steal"
            spread >= 0.12 -> "good"
            spread >= -0.08 -> "fair"
            spread >= -0.30 -> "high"
            else -> "avoid"
        }
        return Appraisal(raw, adjusted, allIn, spread, verdict)
    }

    fun nameQuery(title: String): String {
        val stop = setOf(
            "pokemon", "pokémon", "tcg", "card", "cards", "holo", "rare", "nm", "lp",
            "psa", "bgs", "cgc", "english", "japanese", "raw", "mint",
        )
        val tokens = title.lowercase().replace(Regex("https?://\\S+"), " ")
            .replace(Regex("\\$[0-9,.]+"), " ")
            .replace(Regex("[^a-z0-9'\\s-]"), " ")
            .split(Regex("\\s+"))
            .filter { it.isNotBlank() && it !in stop && !it.all(Char::isDigit) }
        return tokens.take(4).joinToString(" ").ifBlank { "pokemon" }
    }
}
